### DISCUSSION TITLE
Calculate free space error when upload object

### DISCUSSION THREAD

**Discussion by invent-lwj** [minio/minio#12371] [Q&A] [answered]
I check code found that only check all disks total free space greater than object size, if true, choose a pool. But it's will store parity besides data. So should it check data and parity size?

In `getServerPoolsAvailableSpace` function:
```
	for i, zinfo := range storageInfos {
		var available uint64
		var total uint64
		for _, disk := range zinfo.Disks { //计算pool内所有disk容量之和
			total += disk.TotalSpace
			available += disk.TotalSpace - disk.UsedSpace
		}
		// Make sure we can fit "size" on to the disk without getting above the diskFillFraction
		if available < uint64(size) {
			available = 0
		}
		if available > 0 {
			// How much will be left after adding the file.
			available -= -uint64(size)

			// wantLeft is how much space there at least must be left.
			wantLeft := uint64(float64(total) * (1.0 - FractionFraction))
			if available <= wantLeft {
				available = 0
			}
		}
		serverPools[i] = poolAvailableSpace{
			Index:     i,
			Available: available,
		}
	}
```

**Comment by harshavardhana** [comment, marked as answer]
> 	for i, zinfo := range storageInfos {
> 		var available uint64
> 		var total uint64
> 		for _, disk := range zinfo.Disks { //计算pool内所有disk容量之和
> 			total += disk.TotalSpace
> 			available += disk.TotalSpace - disk.UsedSpace
> 		}
> 		// Make sure we can fit "size" on to the disk without getting above the diskFillFraction
> 		if available < uint64(size) {
> 			available = 0
> 		}
> 		if available > 0 {
> 			// How much will be left after adding the file.
> 			available -= -uint64(size)
> 
> 			// wantLeft is how much space there at least must be left.
> 			wantLeft := uint64(float64(total) * (1.0 - FractionFraction))
> 			if available <= wantLeft {
> 				available = 0
> 			}
> 		}
> 		serverPools[i] = poolAvailableSpace{
> 			Index:     i,
> 			Available: available,
> 		}
> 	}

> I check code found that only check all disks total free space greater than object size, if true, choose a pool. But it's will store parity besides data. So should it check data and parity size?

The size does not mean size it is chosen to 2*size look at the top-level caller of the function - the actual function is getPoolIdx(). 

```
// getPoolIdx returns the found previous object and its corresponding pool idx,
// if none are found falls back to most available space pool.
func (z *erasureServerPools) getPoolIdx(ctx context.Context, bucket, object string, size int64) (idx int, err error) {
        idx, err = z.getPoolIdxExisting(ctx, bucket, object)
        if err != nil && !isErrObjectNotFound(err) {
                return idx, err
        }

        if isErrObjectNotFound(err) {
                // We multiply the size by 2 to account for erasure coding.
                idx = z.getAvailablePoolIdx(ctx, size*2)
                if idx < 0 {
                        return -1, toObjectErr(errDiskFull)
                }
        }

        return idx, nil
}
```
